source/dofs/number_cache.cc

```diff
@@ -136,95 +125,8 @@ namespace internal
-                boost::archive::binary_oarchive archive(out);
-                archive << locally_owned_dofs;
-                const std::string &s = out.str();
-                my_data.reserve(s.size());
-                my_data.assign(s.begin(), s.end());
-#  endif
-              }
-
-              // determine maximum size of IndexSet
-              const unsigned int max_size =
-                Utilities::MPI::max(my_data.size(), mpi_communicator);
-
-              // as the MPI_Allgather call will be reading max_size elements,
-              // and as this may be past the end of my_data, we need to increase
-              // the size of the local buffer. This is filled with zeros.
-              my_data.resize(max_size);
-
-              std::vector<char> buffer(max_size * n_procs);
-              const int         ierr = MPI_Allgather(my_data.data(),
-                                             max_size,
-                                             MPI_BYTE,
-                                             buffer.data(),
-                                             max_size,
-                                             MPI_BYTE,
-                                             mpi_communicator);
-              AssertThrowMPI(ierr);
-
-              for (unsigned int i = 0; i < n_procs; ++i)
-                if (i == Utilities::MPI::this_mpi_process(mpi_communicator))
-                  locally_owned_dofs_per_processor[i] = locally_owned_dofs;
-                else
-                  {
-                    // copy the data previously received into a stringstream
-                    // object and then read the IndexSet from it
-                    std::string decompressed_buffer;
-
-                    // first decompress the buffer
-                    {
-#  ifdef DEAL_II_WITH_ZLIB
-
-                      boost::iostreams::filtering_ostream decompressing_stream;
-                      decompressing_stream.push(
-                        boost::iostreams::gzip_decompressor());
-                      decompressing_stream.push(
-                        boost::iostreams::back_inserter(decompressed_buffer));
-
-                      decompressing_stream.write(&buffer[i * max_size],
-                                                 max_size);
-#  else
-                      decompressed_buffer.assign(&buffer[i * max_size],
-                                                 max_size);
-#  endif
-                    }
-
-                    // then restore the object from the buffer
-                    std::istringstream              in(decompressed_buffer);
-                    boost::archive::binary_iarchive archive(in);
-
-                    archive >> locally_owned_dofs_per_processor[i];
-                  }
-            }
-#endif
-          return locally_owned_dofs_per_processor;
+          return Utilities::MPI::all_gather(mpi_communicator,
+                                            locally_owned_dofs);
```

Comment:
here you use this function, but I think we can remove the function altogether - this class is mostly internal anyway, I think.